File: src/games/marbles.rs
use rand::Rng;

#[derive(Debug, Clone)]
pub struct MarblesMove(pub u32 /* amount to remove */);

#[derive(Clone)]
pub struct MarblesState {
    pub pile_amt: u32, 
    pub turn: bool, // true if it's player's turn, false is computer's turn
    pub moves: Vec<MarblesMove> // store all moves made in the game
}

impl MarblesState {
    pub fn new(pile_amt: u32, turn: bool) -> Self {
        Self { pile_amt, turn, moves: Vec::new() }
    }

    pub fn gen() -> Self {
        let mut rng = rand::thread_rng();
        let pile_amt = rng.gen_range(10..=20);
        let turn = true;
        Self::new(pile_amt, turn)
    }

    pub fn is_game_over(&self) -> bool {
        self.pile_amt == 0
    }

    // note: refactor to return a Result
    pub fn apply_move(&mut self, action: MarblesMove) -> bool {
        let MarblesMove(amount) = action;
        if amount <= 0 || amount > 3 {
            return false;
        }
        if amount > self.pile_amt {
            return false;
        }
        self.pile_amt -= amount;
        self.turn = !self.turn;
        self.moves.push(action);
        true
    }

    pub fn grundy_value(&self) -> u32 {
        self.pile_amt % 4
    }

    pub fn optimal_move(&self) -> MarblesMove {
        let grundy = self.grundy_value();
        if grundy == 0 {
            // every move loses, pick an arbitrary move
            MarblesMove(1)
        } else {
            MarblesMove(grundy)
        }
    }

    pub fn undo_move(&mut self) {
        if let Some(MarblesMove(amount)) = self.moves.pop() {
            self.pile_amt += amount;
            self.turn = !self.turn;
        }
    }
}

File: games/marbles.ts
import { MarblesMove } from "@/types/marbles";

/** Represents the state of the Marbles game */
export class MarblesState {
    public marbles: number;
    public turn: boolean; // true if it's the player's turn, false if it's the computer's turn
    public moves: MarblesMove[];
    public maxMarblesPerTurn: number;

    constructor(marbles: number, turn: boolean, maxMarblesPerTurn: number = 3, moves: MarblesMove[] = []) {
        this.marbles = marbles;
        this.turn = turn;
        this.maxMarblesPerTurn = maxMarblesPerTurn;
        this.moves = moves;
    }

    /** Clones the game state */
    clone(): MarblesState {
        return new MarblesState(this.marbles, this.turn, this.maxMarblesPerTurn, [...this.moves]);
    }

    /** Generates a random game state */
    static gen(): MarblesState {
        const marbles = Math.floor(Math.random() * (20 - 10 + 1)) + 10; // Random number between 10 and 20
        const turn = false;
        return new MarblesState(marbles, turn);
    }

    /** Checks if the game is over */
    isGameOver(): boolean {
        return this.marbles <= 0;
    }

    /** Applies a move to the game state */
    applyMove(action: MarblesMove): boolean {
        const { amount } = action;

        if (amount > this.maxMarblesPerTurn || amount <= 0 || this.marbles < amount) {
            return false;
        }

        this.marbles -= amount;
        this.turn = !this.turn;
        this.moves.push(action);
        return true;
    }

    /** Calculates the Grundy value of the current state */
    grundyValue(): number {
        return this.marbles % (this.maxMarblesPerTurn + 1);
    }

    /** Calculates the optimal move for the current state */
    optimalMove(): MarblesMove {
        const grundy = this.grundyValue();

        if (grundy === 0) {
            // Every move loses, pick an arbitrary move
            return { amount: 1 };
        }

        return { amount: grundy };
    }

    /** Undoes the last move */
    undoMove(): void {
        const lastMove = this.moves.pop();
        if (lastMove) {
            this.marbles += lastMove.amount;
            this.turn = !this.turn;
        }
    }
}
